### DISCUSSION TITLE
issue upgrading typer 0.9.0 to 0.12.0 (typer directory in site packages appears to contain only typer-cli)

### DISCUSSION THREAD

**Discussion by brianmedigate** [fastapi/typer#784] [Questions] [answered]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the Typer documentation, with the integrated search.
- [X] I already searched in Google "How to X in Typer" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to Typer but to [Click](https://github.com/pallets/click).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
$ python3.8 -m virtualenv venv
$ . venv/bin/activate
$ pip install typer==0.9.0
$ pip install typer==0.12.0
$ ll venv/lib/python3.8/site-packages/typer/
total 36
drwxrwxr-x  3 brianm brianm 4096 Mar 31 17:52 ./
drwxrwxr-x 28 brianm brianm 4096 Mar 31 17:52 ../
-rw-rw-r--  1 brianm brianm 9149 Mar 31 17:52 cli.py
-rw-rw-r--  1 brianm brianm 6689 Mar 31 17:52 _completion_classes.py
-rw-rw-r--  1 brianm brianm   30 Mar 31 17:52 __main__.py
drwxrwxr-x  2 brianm brianm 4096 Mar 31 17:52 __pycache__/
```

### Description

After upgrading typer 0.9.0 to 0.12.0 in an existing virtual environment, the `typer` directory in site packages doesn't contain the expected code of typer. It looks like it only contains typer-cli or something like that. I'm guessing it has something to do with the recent packaging changes in https://github.com/tiangolo/typer/pull/780.

### Operating System

Linux

### Operating System Details

Ubuntu 22.04.3

### Typer Version

0.12.0

### Python Version

3.8.18

**Comment by brianmedigate** [comment]
A workaround in case it helps anyone else, uninstall `typer-slim` and then reinstall `typer`:
```sh
$ pip uninstall typer-slim
$ pip install typer==0.12.0
```
Uninstalling just `typer` doesn't solve it.

Or of course just recreate the virtual environment

**Reply by tiangolo** [reply]
Thanks for the report @brianmedigate! I made it an issue https://github.com/tiangolo/typer/issues/790 and I'm working on it. :nerd_face:

**Reply by tiangolo** [reply, marked as answer]
This was fixed in https://github.com/tiangolo/typer/pull/791, released in Typer `0.12.1` :rocket: :tada:

**Comment by lone17** [comment]
I had the same problem. After further inspection, I think it is indeed due to the recent packaging changes. Here is what I found:

### Problem
After the latest packing changes, `typer-slim` will write to `site-packages/typer/` instead of `typer` as in older versions.

During the upgrade from a lower version, the dependencies will be installed before the old version is uninstalled (from pip's [source code](https://github.com/pypa/pip/blob/main/src/pip/_internal/req/__init__.py#L54-L92)). So what happens are:
1. `typer-slim` is installed, writing to  `site-packages/typer/`. (also `typer-cli` is installed but that's irrelevant)
2. Old `typer` is uninstalled, removing files from  `site-packages/typer/`, while still leaving 2 files added by `typer-slim` (`cli.py` and `__main__.py`).
3. New `typer` is installed and nothing is written to  `site-packages/typer`.

Edit: `poetry` also follows a similar upgrade procedure, as mentioned by @HansAarneLiblik.

### Consequences
This leads to a few things:
- Upgrading from an older version will break the package.
  - So users need to uninstall `typer` before installing `0.12.0`.
  - Or if you've already upgraded the package, then you need to uninstall `typer-slim` as mentioned by @brianmedigate then install the new version.
- `pip uninstall typer` in `0.12.0` will not work as expected because `typer-slim` still exists. So you can still `import typer` but cannot `pip show typer`.
- `pip uninstall typer-slim` in `0.12.0` will break `typer`. This means `pip show typer` works but `import typer` doesn't.

These issues would create quite confusion as someone upgrading from an older version will encounter it while someone doing a fresh install will not.

### Solution
I suppose the solution would be changing pack to the previous packaging and converting `typer-slim` to an optional dependency `typer[slim]` instead.

**Reply by sizur** [reply]
Thank you!  This helped mitigate the breaking change.  Opening issues on github is not welcome even.  I fear we'll have to make some decisions against some libraries I love.

**Reply by lone17** [reply]
I'm sure the maintainer will fix this in the next release so don't worry :)

**Reply by tiangolo** [reply]
Awesome investigation, this helps a lot, thank you @lone17! :bow: 

I'm working on it. :nerd_face: 

I created an issue to track it here: https://github.com/tiangolo/typer/issues/790

**Reply by tiangolo** [reply]
This was fixed in https://github.com/tiangolo/typer/pull/791, released in Typer `0.12.1` :rocket: :tada: 

Could you help me confirm it solves it?

**Comment by HansAarneLiblik** [comment]
I also witnessed this when using `poetry` `v1.8.2` and upgrading `typer` `0.11.1 -> 0.12.0`, and in the logs I see

```
  - Installing typer-slim (0.12.0): Installing...
Installing <venv_path>/lib/python3.12/site-packages/typer/__init__.py over existing file
Installing <venv_path>/lib/python3.12/site-packages/typer/_completion_classes.py over existing file
Installing <venv_path>/lib/python3.12/site-packages/typer/_completion_shared.py over existing file
Installing <venv_path>/lib/python3.12/site-packages/typer/_typing.py over existing file
Installing <venv_path>/lib/python3.12/site-packages/typer/colors.py over existing file
Installing <venv_path>/lib/python3.12/site-packages/typer/completion.py over existing file
Installing <venv_path>/lib/python3.12/site-packages/typer/core.py over existing file
Installing <venv_path>/lib/python3.12/site-packages/typer/main.py over existing file
Installing <venv_path>/lib/python3.12/site-packages/typer/models.py over existing file
Installing<venv_path>/lib/python3.12/site-packages/typer/params.py over existing file
Installing <venv_path>/lib/python3.12/site-packages/typer/py.typed over existing file
Installing <venv_path>/lib/python3.12/site-packages/typer/rich_utils.py over existing file
Installing <venv_path>/lib/python3.12/site-packages/typer/testing.py over existing file
  - Installing typer-slim (0.12.0)
  ...
  - Installing typer-cli (0.12.0)
  ...
  - Updating typer (0.11.1 -> 0.12.0)
  ```
  
  which made it clear that something is messed up

**Reply by tiangolo** [reply]
This was fixed in https://github.com/tiangolo/typer/pull/791, released in Typer `0.12.1` :rocket: :tada: 

Could you help me confirm it solves it?

**Reply by HansAarneLiblik** [reply]
ISTM that upgrading from 0.10.0 to 0.12.1 worked!

Thank you!

**Comment by brianmedigate** [comment]
@tiangolo I think this might require your attention

**Comment by FCamborda** [comment]
I posted a related issue: https://github.com/tiangolo/typer/discussions/786
This hints to me that there's a new general packaging problem in this project.

**Comment by jansila** [comment]
Upgrading to `0.12.0` also messes up imports , just fyi if someone comes across this issue
```
>>> import typer
>>> typer.Argument()
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
AttributeError: module 'typer' has no attribute 'Argument'
```

**Reply by HansAarneLiblik** [reply]
Yes, this is the issue. It's not that the imports are "messed up", it's because `typer-slim` deleted the actual Python files

**Reply by lone17** [reply]
It's more than just messing up, your `typer` package is actually empty now 😂

**Reply by tiangolo** [reply]
This was fixed in https://github.com/tiangolo/typer/pull/791, released in Typer `0.12.1` :rocket: :tada: 

Could you help me confirm it solves it?

**Reply by corneliusroemer** [reply]
No, I still get a similar error with 0.12.3:

```
Traceback (most recent call last):
  File "/opt/homebrew/Caskroom/miniforge/base/envs/py11/bin/nwkfmt", line 3, in <module>
    from nwkfmt.cli import entry_point
  File "/Users/corneliusromer/code/nwk-formatter/nwkfmt/cli.py", line 7, in <module>
    app = typer.Typer(context_settings={"help_option_names": ["-h", "--help"]})
          ^^^^^^^^^^^
AttributeError: module 'typer' has no attribute 'Typer'
```

When reverting to 0.11.1 there's no issue.

Filed a discussion here: https://github.com/tiangolo/typer/discussions/799#discussion-6500401
